### PR TITLE
Add support for encrypted Realms

### PR DESCRIPTION
**Pull request by clementetb**
With this PR we add support for encrypted Realms. Realms are encrypted with a 64-byte key, it can be passed into the configuration via the configuration builder. The SDK will throw an `IllegalArgumentException` on any wrong-sized key.

Example:

```
val encryptedRealmConfig = RealmConfiguration.Builder()
    .encryptionKey(encryptionKey)
    .build()
    
val encryptedRealm = Realm(encryptedRealmConfig)
```

### CODE REVIEW

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 1]
```diff
@@ -121,6 +122,11 @@ public class RealmConfiguration private constructor(
      */
     public val deleteRealmIfMigrationNeeded: Boolean
 
+    /**
+     * 64 byte key used to encrypt and decrypt the Realm file.
+     */
+    public val encryptionKey get() = RealmInterop.realm_config_get_encryption_key(nativeConfig)
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:128]: We should probably make the return type explicit here for better discoverability.

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 2]
```diff
@@ -339,6 +351,12 @@ public class RealmConfiguration private constructor(
         fun schemaVersion(schemaVersion: Long) =
             apply { this.schemaVersion = validateSchemaVersion(schemaVersion) }
 
+        /**
+         * Sets the 64 byte key used to encrypt and decrypt the Realm file.
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:355]: We should probably expand this description a little bit with advice on how to do this best, see e.g https://github.com/realm/realm-core/issues/4683.

Maybe something like:
```
Sets a 64 byte key that will be used to encrypt and decrypt the Realm file. 

It is important that this key is created and stored securely. See [this link](https://docs.mongodb.com/realm/sdk/android/advanced-guides/encryption/) for suggestions on how to do that.
```

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 3]
```diff
@@ -386,5 +405,12 @@ public class RealmConfiguration private constructor(
             }
             return schemaVersion
         }
+
+        private fun validateEncryptionKey(encryptionKey: ByteArray): ByteArray {
+            if (encryptionKey.size < Realm.ENCRYPTION_KEY_LENGTH) {
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:410]: We should check for equality, not just lesser than?

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+import io.realm.RealmConfiguration
+import io.realm.util.PlatformUtils
+import test.Sample
+import kotlin.random.Random
+import kotlin.test.AfterTest
+import kotlin.test.BeforeTest
+import kotlin.test.Test
+import kotlin.test.assertFails
+
+class EncryptionTests {
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:29]: Maybe put some doc explaining what this class exactly tests compared to RealmConfigurationTests?

File: test/src/androidTest/kotlin/io/realm/shared/RealmConfigurationTests.kt [hunk 5]
```diff
@@ -241,4 +244,33 @@ class RealmConfigurationTests {
...
+            .encryptionKey(encryptionKey)
+            .build()
+
+        // Validate that the key stored in core is the same that the one we provided
+        assertContentEquals(encryptionKey, config.encryptionKey)
+    }
+
+    @Test
+    fun wrongEncryptionKeyThrowsIllegalArgumentException() {
+        val encryptionKey = Random.nextBytes(8)
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/RealmConfigurationTests.kt:268]: Should also test for longer keys

**Review by cmelchior — APPROVED** [review]
Nice 👏 I only had a few minor comments

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+        // Initializes an encrypted Realm
+        Realm(encryptedConf).close()
+
+        // Should be possible to reopen an encrypted Realm
+        Realm(encryptedConf).close()
+    }
+
+    @Test
+    fun openEncryptedRealmWithWrongKey() {
```
Comment (rorbech) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:61]: Do we also need the opposite test? Opening an unencrypted realm with a key.

File: test/src/macosTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 7]
```diff
@@ -0,0 +1 @@
+../../../../../androidTest/kotlin/io/realm/shared/EncryptionTests.kt
```
Comment (rorbech) [inline test/src/macosTest/kotlin/io/realm/shared/EncryptionTests.kt:1]: I thought the whole `shared` directory was symlinked. Would that be possible?

**Review by rorbech — APPROVED** [review]
Same. Only minor comments

File: test/src/macosTest/kotlin/io/realm/shared/EncryptionTests.kt (hunk 7 above)
Comment (clementetb) [reply test/src/macosTest/kotlin/io/realm/shared/EncryptionTests.kt:1]: Yes it is, I am going to address it on another PR

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 8]
```diff
@@ -121,6 +122,11 @@ public class RealmConfiguration private constructor(
      */
     public val deleteRealmIfMigrationNeeded: Boolean
 
+    /**
+     * 64 byte key used to encrypt and decrypt the Realm file.
```
[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:126]: This doesn't seem to have been addressed?

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 9]
```diff
@@ -339,6 +351,17 @@ public class RealmConfiguration private constructor(
         fun schemaVersion(schemaVersion: Long) =
             apply { this.schemaVersion = validateSchemaVersion(schemaVersion) }
 
+        /**
+         * Sets the 64 byte key used to encrypt and decrypt the Realm file. Not setting an encryption key would
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:355]:
```suggestion
         * Sets the 64 byte key used to encrypt and decrypt the Realm file. If no key is provided the Realm file 
```

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 10]
```diff
@@ -339,6 +351,17 @@ public class RealmConfiguration private constructor(
...
+         * default to an unencrypted Realm.
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:356]:
```suggestion
         * will be unencrypted.
```

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 11]
```diff
@@ -339,6 +351,17 @@ public class RealmConfiguration private constructor(
...
+         *
+         * It is important that this key is created and stored securely. See [this link](https://docs.mongodb.com/realm/sdk/android/advanced-guides/encryption/) for suggestions on how to do that.
+         *
+         * @param encryptionKey 64-byte key encryption key.
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:360]:
```suggestion
         * @param encryptionKey 64-byte encryption key.
```

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 12]
```diff
@@ -386,5 +410,12 @@ public class RealmConfiguration private constructor(
             }
             return schemaVersion
         }
+
+        private fun validateEncryptionKey(encryptionKey: ByteArray): ByteArray {
+            if (encryptionKey.size != Realm.ENCRYPTION_KEY_LENGTH) {
+                throw IllegalArgumentException("The provided key must be ${Realm.ENCRYPTION_KEY_LENGTH} bytes.")
```
Comment (cmelchior) [inline packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:416]:
```suggestion
                throw IllegalArgumentException("The provided key must be ${Realm.ENCRYPTION_KEY_LENGTH} bytes. The provided key was ${encryptionKey.size} bytes.")
```

We should always try to give as much information as possible in these crash messages. It becomes much easier to debug.

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+                .path("$tmpDir/default.realm")
+                .encryptionKey(randomKey)
+                .build()
+
+            Realm(conf)
+        }
+    }
+
+    @Test
+    fun openUnencryptedRealmWithWrongKey() {
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:100]:
```suggestion
    fun openUnencryptedRealmWithEncryptionKey() {
```

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+            .Builder(
+                path = "$tmpDir/default.realm",
+                schema = setOf(Sample::class)
+            )
+            .build()
+        Realm(unencryptedConf).close()
+
+        // Assert fails opening with encryption key
+        val randomKey = Random.nextBytes(64)
+        assertFails {
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:112]: We should probably use `assertFailsWith(XXX::class)` here to provide more context about exactly how it fails.

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+            val conf = RealmConfiguration.Builder(schema = setOf(Sample::class))
+                .path("$tmpDir/default.realm")
+                .build()
+            Realm(conf)
+        }
+
+        // Assert fails with wrong encryption key
+        val randomKey = Random.nextBytes(64)
+        assertFails {
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:89]: We should probably use assertFailsWith(XXX::class) here to provide more context about exactly how it fails.

File: test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt [hunk 16]
```diff
@@ -0,0 +1,121 @@
...
+            .Builder(
+                path = "$tmpDir/default.realm",
+                schema = setOf(Sample::class)
+            )
+            .encryptionKey(actualKey)
+            .build()
+        Realm(encryptedConf).close()
+
+        // Assert fails with no encryption key
+        assertFails {
```
Comment (cmelchior) [inline test/src/androidTest/kotlin/io/realm/shared/EncryptionTests.kt:80]: We should probably use assertFailsWith(XXX::class) here to provide more context about exactly how it fails.

**Review by cmelchior — APPROVED** [review]
A few minor comments around the tests and some docs, otherwise 🚀

**Pull request merged**